### ISSUE TITLE
(inactive) OptiLCMS

### ISSUE THREAD

**Issue by Zhiqiang-PANG** [Bioconductor/Contributions#2210]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/xia-lab/OptiLCMS

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by LiNk-NY** [comment]
Hi Zhiqiang, @Zhiqiang-PANG 

Please subscribe to the bioc-devel mailing list with the same email as listed in the package `DESCRIPTION`:

```
* Checking for bioc-devel mailing list subscription...
    * ERROR: Maintainer must subscribe to the bioc-devel mailing list.
      Subscribe here: https://stat.ethz.ch/mailman/listinfo/bioc-devel
```

PS. Any reason why one of the 10 tasks wasn't checked off above?

**Comment by Zhiqiang-PANG** [comment]
Hi LiNk-NY, @LiNk-NY 

I have subscribed the mailing list. Thanks for reminding. For the task I did not check off, I think my R package is about metabolomics data processing, rather genomics. If you have any other comment, please let me know immediately. Thank you so much!
Zhiqiang

**Comment by LiNk-NY** [comment]
Hi Zhiqiang, @Zhiqiang-PANG 
Can you speak to the timeout? 
Best,
Marcel

**Comment by Zhiqiang-PANG** [comment]
Hi Marcel, @LiNk-NY 
It looks the MacOS platform running the vignette for too long. But I run the steps from my MacOS (8GB RAM + 6 CPUs), it only takes 10.2min. I will try to improve some codes inside the vignette. Do you have any other advice?
Thanks,
Zhiqiang

**Comment by LiNk-NY** [comment]
Hi Zhiqiang, @Zhiqiang-PANG 

Thank you for the response.

We recommend that you keep the datasets and examples minimal so that you don't hit this timeout limit. 
I will take a look at your package in the coming days. 

Best regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi Zhiqiang, @Zhiqiang-PANG 
Do you have any updates? 
Best,
Marcel

**Comment by Zhiqiang-PANG** [comment]
Hi @LiNk-NY 
I don't have any other updates since last commits, because the example has already been minimized in the last commit and the time-out issue has also been resolved.
Thanks,
Zhiqiang

**Comment by LiNk-NY** [comment]
Hi ZhiQiang, @Zhiqiang-PANG 
Can you bump the version for another build? Thank you.

**Comment by LiNk-NY** [comment]
Hi ZhiQiang, @Zhiqiang-PANG 
Please bump the version of the package to keep the issue going.
Otherwise, it will be closed for inactivity. 
Best,
Marcel

**Comment by LiNk-NY** [comment]
Hi ZhiQiang, @Zhiqiang-PANG 
It seems that the timeout it back. Please resolve. 
I am currently reviewing the package. 
Best regards,
Marcel

**Comment by Zhiqiang-PANG** [comment]
Hi Marcel, @LiNk-NY 
I am fixing now. Thanks.

Bests,
Zhiqiang

**Comment by LiNk-NY** [comment]
Hi Zhiqiang, @Zhiqiang-PANG 

Thank you for your submission. 
Please see the review below. 

Best regards,
Marcel

---

# OptiLCMS #2210

The package will need a lot of work before it can be accepted. Unfortunately,
given the lack of adherence to standard Bioconductor practices, the code can
be extremely fragile and unreliable.

## DESCRIPTION

* If you only need one funciton from `MSnbase`, please put the package in the
Imports field.
* Minor: Keep the `Description` field limited to 80 column width.

## NAMESPACE

* Avoid using `.` delimiters in function names. These are meant to indicate an
S3 method dispatch. Use `camelCase` instead.

## vignettes/

* Consider limiting the text to 80 column-width. This helps reviewers and
even maintenance.
* Remove the Google Drive link in the vignette. If you'd like to make the
source tarball available, create a "release" on your GitHub page (though this
is not the usual way to distribute the package).
* Remove `eval=TRUE` flags from code chunks. This is the default.
* Minimize the number of `eval=FALSE` code blocks. If the data is too big,
consider making it smaller.
* Consider breaking down the `running.plan` function call in to several
smaller steps. It looks quite complicated and not very user friendly.
* Avoid using semicolons (`;`).

## R/

* Avoid double logical evaluation, e.g., (`if (TRUE) TRUE else FALSE` in
`continuousPtsAbove*` etc.)
* Use consistent delimiters in a error message (e.g., in `which.colMax` :
`'x'`)
* Use `as.double` without checking to see if a matrix `is.double` because you
will coerce anyway. This applies to other coercions where you're checking
first before coercing. Checking is not necessary.
* Consider using a helper function to check the inputs of the `binYonX` and
other related functions.
* Your comment in `binYonX` and `fast.write.csv` does not apply because your
package will not be available for older R versions only versions 4.1 and above
(once in Bioconductor; if accepted).
* Check for `NA` in `x` earlier in the function (especially if it is an error).
* Avoid writing files without the user's consent (e.g., `write.table` in
`MessageOutput` and others).
* Remove any `.onAttach` messages, this is against Bioconductor standards of
practice.
* Avoid empty validity checks (e.g., in the `controller` and `ResumeHistory`
classes) and add a validity function to ensure the proper class structure.
* Does the `ResumingePlan` class have a typo in it?
* I do not see an `OptiLCMS` class defined in the `R/OptiLCMS_Class.R` file.
* Remove `;` from the code.
* Remove excitement from error messages (i.e., exclamation mark `!`) and
'Don't worry' / 'ADVICE' directives (in `PerformParamsOptimization`).
* Default the `ncore` value to a single core.
* Remove or fix any commented code in R files.
* Avoid nesting `ifelse` functions and use a table
https://stackoverflow.com/a/39217596
* Avoid the use of `nSlaves` / `SlaveCluster` terms.
* Do not use `@` to access slots and use a formal interface detached from the
slotNames in the event that these need to change (e.g. in `calcCV`).
* Avoid using the copy and append method within `for` loops as it can cause
major slow downs (e.g., in `getRGTVValues`)
* Add authors to `DESCRIPTION` for any code that you are porting into your
package. Note that it is common courtesy to ask the original authors for
permission to port their code into your package.
* Avoid manipulating the working directory inside a function. This is not
standard practice. See `GeneratePeakList`.
* Do not hardcode file paths in a function (`CachePathCorrection`).
* Make it possible for users to turn off the excess of messages in
`FastRunningShow_auto` with a `verbose = FALSE` option at the top level
function.
* Is there a better way to `SetPeakParam` without almost 2000 lines of
conditional code?

## tests/

* The maintainers will need a large amount of tests to make sure that all
the functionality in the package is reliable and robust.
* Currently there are no tests in the package.

**Comment by molesong** [comment]
Dear Zhiqiang, @Zhiqiang-PANG

Thank you for your contribution.
I have a problem when I installed the package.Please check it.Thank you very much!

> D:/Program Files/rtools40/mingw64/bin/g++ -std=gnu++11 -shared -s -static-libgcc -o OptiLCMS.dll tmp.def massifquant/xcms_massifquant.o massifquant/TrMgr.o massifquant/Tracker.o massifquant/SegProc.o massifquant/DataKeeper.o massifquant/OpOverload.o obiwarp/mat.o obiwarp/vec.o obiwarp/xcms_dynprog.o obiwarp/xcms_lmat.o xcms_obiwarp.o mzROI.o util.o xcms_binners.o init.o -lgfortran -lm -lquadmath -LD:/PROGRA~1/R/R-41~1.2/bin/x64 -lR
>installing to D:/Program Files/R/R-4.1.2/library/00LOCK-OptiLCMS/00new/OptiLCMS/libs/x64
** R
Warning: unable to re-encode 'Optimize_params.R' line 1839
Warning in readLines(f, warn = FALSE) :
  invalid input found on input connection '
>C:/Users/31219/AppData/Local/Temp/RtmpMLDchH/R.INSTALLcd059246acc/OptiLCMS/R/Spectra_Utils.R'
Error in parse(outFile) : 
  C:/Users/31219/AppData/Local/Temp/RtmpMLDchH/R.INSTALLcd059246acc/OptiLCMS/R/Spectra_visulization.R:1541:0: unexpected end of input
1539: 
1540: 
     ^
>ERROR: unable to collate and parse R files for package 'OptiLCMS'
>* removing 'D:/Program Files/R/R-4.1.2/library/OptiLCMS'
>Warning in install.packages :
>  installation of package ‘D:/All_download/OptiLCMS_0.99.7.tar.gz’ had non-zero exit status
